mmseg/models/backbones/twins.py

```diff
@@ -0,0 +1,899 @@
+                 qkv_bias=False,
+                 qk_scale=None,
+                 attn_drop=0.,
+                 proj_drop=0.,
+                 ws=1,
+                 sr_ratio=1.0):
+        """ws 1 for stand attention."""
+        super(GroupAttention, self).__init__()
+        assert dim % num_heads == 0, f'dim {dim} should be divided by ' \
+                                     f'num_heads {num_heads}.'
+
+        self.dim = dim
+        self.num_heads = num_heads
+        head_dim = dim // num_heads
+        self.scale = qk_scale or head_dim**-0.5
+
+        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+        self.attn_drop = nn.Dropout(attn_drop)
+        self.proj = nn.Linear(dim, dim)
+        self.proj_drop = nn.Dropout(proj_drop)
+        self.ws = ws
+
+    def forward(self, x, H, W):
+        B, N, C = x.shape
+        x = x.view(B, H, W, C)
+        pad_l = pad_t = 0
+        pad_r = (self.ws - W % self.ws) % self.ws
+        pad_b = (self.ws - H % self.ws) % self.ws
+        x = F.pad(x, (0, 0, pad_l, pad_r, pad_t, pad_b))
+        _, Hp, Wp, _ = x.shape
+        _h, _w = Hp // self.ws, Wp // self.ws
+        mask = torch.zeros((1, Hp, Wp), device=x.device)
+        mask[:, -pad_b:, :].fill_(1)
+        mask[:, :, -pad_r:].fill_(1)
+
+        x = x.reshape(B, _h, self.ws, _w, self.ws,
+                      C).transpose(2, 3)  # B, _h, _w, ws, ws, C
+        mask = mask.reshape(1, _h, self.ws, _w,
+                            self.ws).transpose(2, 3).reshape(
+                                1, _h * _w, self.ws * self.ws)
+        attn_mask = mask.unsqueeze(2) - mask.unsqueeze(
+            3)  # 1, _h*_w, ws*ws, ws*ws
+        attn_mask = attn_mask.masked_fill(attn_mask != 0,
+                                          float(-1000.0)).masked_fill(
+                                              attn_mask == 0, float(0.0))
+        qkv = self.qkv(x).reshape(B, _h * _w, self.ws * self.ws, 3,
+                                  self.num_heads, C // self.num_heads).permute(
+                                      3, 0, 1, 4, 2,
+                                      5)  # n_h, B, _w*_h, nhead, ws*ws, dim
+        q, k, v = qkv[0], qkv[1], qkv[2]  # B, _h*_w, n_head, ws*ws, dim_head
+        attn = (q @ k.transpose(
+            -2, -1)) * self.scale  # B, _h*_w, n_head, ws*ws, ws*ws
+        attn = attn + attn_mask.unsqueeze(2)
+        attn = attn.softmax(dim=-1)
+        attn = self.attn_drop(
+            attn)  # attn @v ->  B, _h*_w, n_head, ws*ws, dim_head
+        attn = (attn @ v).transpose(2, 3).reshape(B, _h, _w, self.ws, self.ws,
+                                                  C)
+        x = attn.transpose(2, 3).reshape(B, _h * self.ws, _w * self.ws, C)
+        if pad_r > 0 or pad_b > 0:
+            x = x[:, :H, :W, :].contiguous()
+        x = x.reshape(B, N, C)
+        x = self.proj(x)
+        x = self.proj_drop(x)
+        return x
+
+
+class SpatialReductionAttention(MultiheadAttention):
```

Comment:
Can inherit [EfficientMultiheadAttention](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/mit.py#L91) and name it to GlobalMSA.